MetaMask/metamask-extension
# Add privacy snapshot and fail E2E test on new host requested

## **Description**
Adds a flag to generate a privacy snapshot file to the run-e2e-test command that can be used to require extra scrutiny on Pull requests that add new domains/hosts to the privacy snapshot. This is achieved in two ways:

1. Without adding the `--update-privacy-snapshot` flag to the e2e test run, the test will fail if a new host is detected being accessed. 
2. In a future PR `.codeowners` will be updated to add an owner for the privacy shapshot file so that when it is changed, it requires review from a subset of engineers. This prevents accidental merge without someone checking off the new host. All new hosts will need to have some sort of privacy toggle associated to allow our users to **opt in** to using the new provider. 

The mechanism for recording these requests is fairly straightforward.

1. The mock-e2e file was modified such that the setupMocking function returns three things. First, it returned the mockedEndpoint. mockedEndpoint is what is returned from the testSpecificMock function that some tests declare. There is no type enforcement on this, nothing guarantees that mockedEndpoint is actually a mockedEndpoint. In fact in some cases we return an array of mockedEndpoints. In some tests we return nothing. It also returns the a new function called getPrivacyReport.
2. getPrivacyReport returns the current privacyReport, which is a `Set` that forces the uniqueness of values. privacyReport is updated by a `.on('request-initiated')` handler on the mockttp instance. This handler adds the host of the request to the privacyReport by calling `.add(request.headers.host)` which will only add the string if it doesn't already exist in the set. getPrivacyReport returns a plain javascript array sorted alphanumerically.
3. in helpers, we call getPrivacyReport after the test suite is ran, and we either update the privacy snapshot if that flag was specified and there are differences in the report and the privacy snapshot, or we fail the test because it violates the snapshot. 


## **Manual testing steps**
1. You can test this PR by removing one of the domains from the privacy-snapshot file and then run e2e tests without the `--update-privacy-snapshot` flag and see that it fails.
2. You can also then test the generation by running it with the flag to see that the domain returns to the snapshot.

## **Related issues**
fixes MetaMask/MetaMask-planning#949

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
